### PR TITLE
Fix autograd gradient normalization for permuted and broadcast tensors

### LINKED ISSUE #68: Autograd add grad error
Hi, I've found a runtime error and made this minimized example:

```rust
#[test]
fn test_add_grad_decreasing_idx() {
    let mut cx = Graph::new();
    let a: GraphTensor<R1<2>> = cx.tensor();
    let a: GraphTensor<R3<1, 1, 2>> = a.expand::<_, LAxes2<0, 1>>();
    let a: GraphTensor<R3<2, 1, 1>> = a.permute::<_, LAxes3<2, 1, 0>>();
    // a.shape.fake = [false, true, true]
    // a.shape.indexes = [0, 2, 1] // note that the idx isn't necessarily increasing (0,1,2)
    let b: GraphTensor<R3<2, 1, 1>> = cx.tensor();
    let weights = vec![a.id, b.id];

    let m: GraphTensor<R3<2, 1, 1>> = a * b;
    let loss: GraphTensor<R0> = m.sum_reduce();
    let _grads = cx.compile(Autograd::new(weights, loss), ());
}
```

The error:
```
thread 'autograd::tests::test_add_grad_decreasing_idx' panicked at /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/tinyvec-1.6.0/src/arrayvec.rs:681:26:
index out of bounds: the len is 0 but the index is 0
stack backtrace:
   ..
   3: tinyvec::arrayvec::ArrayVec<A>::remove
             at /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/tinyvec-1.6.0/src/arrayvec.rs:681:26
   4: luminal::shape::tracker::ShapeTracker::remove_dim
             at /workspaces/coursera-deep-learning-specialization/luminal_original/src/shape/tracker.rs:58:21
   5: luminal_training::autograd::add_grad
             at ./src/autograd.rs:208:13
   6: <luminal_training::autograd::Autograd as luminal::compiler_utils::Compiler>::compile
             at ./src/autograd.rs:113:21
   ..
```

For what I've noticed, the `add_grad` function may assume that the `fwd` `shape.indexes` is always increasing, but in this case, thanks to how the `a.permute` axes were defined (`LAxes3<2, 1, 0>`), the indexes may end up decreasing `[0, 2, 1]` - to note, if the axes were defined as `LAxes3<2, 0, 1>`, the indexes would not decrease and there would be no error. Since this shape also has two fake axes (the fake is `[false, true, true]`), at the last axis iteration, the previous fake axis removal changes the axes length, and so the last iteration goes out of bounds.

Relevant code section:
https://github.com/jafioti/luminal/blob/f61d53f859293d4c5e4d57d264dfc423a98007e0/crates/luminal_training/src/autograd.rs#L202-L211

I'm not sure how to solve this as I'm still learning how the shapes works in overall, but I'll try to experiment more.
I've also linked a PR which shows the example and error.

### PR DESCRIPTION
**Pull request by ppaleja**
## Summary
Fixes #68 - resolves an index out-of-bounds panic in autograd's add_grad function that occurred when computing gradients for tensors with non-monotonic permute indices and fake (broadcasted) dimensions.

### Remarks
I do admit these fixes seem extremely hacky to me. Would appreciate review and even if they work, it might be evidence for the brittleness of ShapeTracker/autograd logic in its current state. 
I also came across another issue with `permute` logic during my debugging, but will make a separate issue for that.

## Commits
The fix is structured across 4 commits for easier cherrypicking:

### 1. Add tests for identifying bug in issue #68 (8c1b5a7)
Reorganized test module structure for better organization
Added comprehensive test cases that reproduce the issue, including:
test_add_grad_decreasing_idx: The original failing case from issue #68
test_add_grad_high_rank: Tests gradient computation with higher-rank tensors
test_add_grad_preserves_unfaked_dim: Tests preservation of non-broadcast dimensions

Status: All new tests fail at this commit (as expected)

### 2. Refactor gradient normalization logic (777d4b2)
Extracted gradient normalization into normalize_grad_for_input helper function
No behavioral changes, pure refactoring
Added unit tests specifically for the normalization logic
Makes the subsequent fixes clearer and more testable

Status: Tests still fail

### 3. Fix add_grad fake-dim removal (Partial fix) (6e898fd)
Core fix: Changed from using logical indices (fwd.shape.indexes) to physical indices when identifying and removing fake dimensions
Now correctly collects fake underlying indices: (0..fwd.shape.len()).filter(|&idx| fwd.shape.fake[idx])
Sorts indices in descending order before removal to prevent index shifting issues
Why this works: Eliminates the mixing of logical/physical index spaces that caused incorrect axis access

Status: Fixes the OOB panic from the original issue, but test_add_grad_preserves_unfaked_dim still fails because rank alignment isn't handled yet

### 4. Fix broadcasted axes in autograd gradient norm (Complete fix) (4057c7e)
Handles the case where incoming gradients have fewer axes than the forward tensor
Inserts missing axes at positions where the forward tensor has fake dimensions before performing undo-permute logic
Uses grad.expand() to ensure gradient shape aligns with forward tensor shape

Remarks: I think this is the hackiest part of the changes, but it does seem necessary to pass `test_add_grad_preserves_unfaked_dim` (which I verified separately is valid logic in pytorch). Unclear if there is a better way.

Status: All tests pass

### CODE REVIEW

**Comment by jafioti** [comment]
great work on this. we've temporarily removed luminal_training, should be back in a month or so
